enhancements/network/on-prem-bgp-networking.md

```diff
@@ -0,0 +1,545 @@
+type OpenStackPlatformSpec struct {
+	// apiLoadBalancer defines how traffic destined to the OpenShift API is
+	// routed to the API servers.
+	// When omitted, this means no opinion and the platform is left to
+	// choose a reasonable default. This default is subject to change over
+	// time.
+	// The current default configuration uses VRRP.
+	//
+	// +optional
+	APILoadBalancer OpenStackAPILoadBalancer `json:"apiLoadBalancer"`
+}
+
+// OpenStackAPILoadBalancer defines how inbound traffic is routed to the API
+// servers.
+type OpenStackAPILoadBalancer struct {
+	// apiLoadBalancerType defines the type of loadbalancer which will be
+	// configured for the API server. Permitted values are `VRRP` and
+	// `BGP`.
+	// When omitted, this means no opinion and the platform is left to
+	// choose a reasonable default. This default is subject to change over
+	// time.
+	// The current default value is `VRRP`.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:="VRRP";"BGP"
+	// +kubebuilder:validation:Required
+	APILoadBalancerType string `json:"type,omitempty"`
+
+	// bgpConfiguration describes the configuration of a BGP load balancer
+	// for the API server. It is only used if apiLoadBalancer is set to
+	// `BGP`.
+	//
+	// +optional
+	BGP *OpenStackAPIBGPConfiguration `json:"bgp,omitempty"`
+}
+```
+
+### Installer
+
+A new [installer](https://github.com/openshift/installer/pull/6440) PR has been proposed, so it'll be possible to:
+- Populate the load-balancer type in the Cluster Infrastructure object, so it can be consumed later (e.g. by MCO).
+- Change the load-balancer type (default remains VRRP) all from `install-config.yaml`.
+
+For now, the API and Ingress VIPs can be any valid IP addresses, but we need to consider two things:
+- There is some [on going work on vSphere](https://github.com/openshift/enhancements/pull/1267) to allow the installer to
+  provide Static IPs for the machines.
+- We might want to create a new parameter `vipSubnet` that will be a Neutron Subnet UUID, so the installer can create the
+  ports in a specific subnet reserved for VIPs. The ports would be disabled, so the network backend doesn't create them but at least their IP be reserved in Neutron
+  so the network administrators can't create ports that overlap with these IPs.
+
+### MCO
+
+The MCO will be responsible for configuring the BGP speakers, based on the API and installer changes described above.
+Further details are described in the following sections.
+
+### Virtual IP addresses and BGP
+
+#### BGP
+
+Earlier we saw why we need dynamic routing instead of stretching L2 networks.
+When a cluster control plane is stretched across multiple L2 domains, the API and Ingress VIPs will be IP addresses that
+aren't part of the machine subnet but live on a loopback device on the host. Since we don't want to stretch L2 networks,
+we need to be able to know on which masters the VIPs are located and also what's the best path to reach it.
```

Comment:
> know on which masters the VIPs are located

That sounds strange, since all masters have the API VIP as a loopback device. We may consider splitting the concern of routing from the concern of balancing, in this paragraph, for the sake of readability.

Consider something like:

> In order to route incoming traffic to the Control plane nodes, we propose using the Border Gateway Protocol. BGP is a routing protocol that is part of the Internet fabric and has already been used in OpenShift (e.g. MetalLB). Each Control plane node would advertise the API and Ingress virtual IPs to the existing BGP routers of the OpenStack cluster.

I would defer implementation details (e.g. the loopback interface) to the "VIPs management" paragraph below

Reply:
:+1: